### ISSUE TITLE
Better experience when session has expired

### ISSUE THREAD

**Issue by adrianbj** [processwire/processwire-requests#384]
### Short description of the enhancement

This is one of those things that has bugged me forever, but it's even worse for clients who don't understand what is happening. If they have the admin open in the browser and at some point their session expires and they go to access items from the main menu that rely on ajax, the spinner pops up, but nothing ever happens because their session is failing in the background. Or, if they have a page edit process open and the session expires before they save it, they will lose their work because they don't have any notification about being logged out.

### Current vs. suggested behavior

I think ideally a modal would popup notifying them that they were logged out and provide the login form so that when they submit that, they will be logged in, the modal will close and they will be returned to exactly where they were and they can save the changes they've made without losing anything.

Getting a bit OT here, but it would also be great if fields could remember content (the way Github and PW forum do) so that if they accidentally close their browser, or their session expires, the content will be re-populated on page load.

**Comment by szabeszg** [comment]
> Getting a bit OT here, but it would also be great if fields could remember content (the way Github and PW forum do) so that if they accidentally close their browser, or their session expires, the content will be re-populated on page load.

I do not think it is off topic by any means as it is also part of the same user experience, which is important.

**Comment by teppokoivula** [comment]
> I do not think it is off topic by any means as it is also part of the same user experience, which is important.

On-topic in the way that "it would be amazing addition". Just wanted to point out that _even without that_ this suggestion would be superb: for me the biggest issue is what Adrian mentioned above about page edit, i.e. I click "save" and get an error, losing all the changes I've made.

(This happens to me a lot — so much that these days I usually manually copy everything in a text editor before hitting save.)

Even something as simple as validating the session with JS and preventing redirect in this situation would be a lifesaver. Having ProcessWire store field values would be nice, but it also sounds like a much bigger feature (esp. if it needs to deal with all the complex, repeatable, custom, and possibly AJAX loaded fields that the page might have) :)

**Comment by jmartsch** [comment]
I wish I could give multiple thumbsup for this issue

**Comment by ryancramerdesign** [comment]
@teppokoivula @adrianbj I don't think there's a way to generically support it for all fields, but for specific fields I think there likely is. For instance, I'm guessing there's got to be a CKEditor plugin that does this, and that's probably what we see in the PW forums CKE instance. We'll have to look into it. 

As for a session timing out during editing, this is not something I've run into or had my clients run into. Is it possible your PHP session settings need adjustment? Any idea how that's happening? I think better than accommodating that undesirable situation would be figure out how to prevent it from happening in the first place. Without knowing what the cause is, I can't say for certain, but maybe a module like ProDevTools UserActivity would be good to integrate into the core, since it keeps an ajax ping going, keeping track of who is editing what and preventing collisions. The ajax ping should prevent even the most aggressive session settings from timing out.

**Comment by adrianbj** [comment]
@ryancramerdesign - I have a long session timeout (using a workaround on all sites because of the Debian GC weirdness), but I don't think it's really that simple - there can be other reasons for a session being ended, but the browser tab is still open.  Sometimes it might even be as simple as someone logging out in one tab but leaving a PW edit screen running in another, then visiting it and not realizing they are not logged in anymore.

Regarding remembering field content if a user accidentally closes the browser etc - I thought this could be handled via the browser's sessionStorage / localStorage with them being cleared on page save. I am envisioning storing json data with page ID, field ID, and content so these values can be used to populate on page render.

**Comment by JanRomero** [comment]
For the original session problem arguably it would be easier to just switch page saves to async requests. No need to add a new feature and make sure everything is compatible, just don’t navigate away. One could then throw up a modal login dialog and continue as if nothing happened.

It would also be nice to have periodic heartbeat requests that extend the session. That should fix most instances of this problem before they happen. Actually, this might be case for a relatively simple third party module?
